### DISCUSSION TITLE
[RFC FS-1068 Discussion] Open static classes

### DISCUSSION THREAD

**Discussion by dsyme** [fsharp/fslang-design#352] [Language and core library RFC discussions]
Discussion thread for https://github.com/fsharp/fslang-design/blob/master/FSharp-5.0/FS-1068-open-type-declaration.md

**Comment by dsyme** [comment]
I've recorded two TBD items for this RFC:

* [x] Resolve how static classes are defined in F#
* [x] Respect the `RequireQualifiedNameAttribute(true)` on a static class, which prevents it being opened, or at least gives a warning, as for F# modules.
* [x] Add testing for AutoOpen attribute

**Comment by dsyme** [comment]
Some discussion here too: https://github.com/fsharp/fslang-suggestions/issues/383

**Comment by dsyme** [comment]
I notice C# lets you do this:

```using static System.Double;```

For some reason I thought it only let you open static classes, but it seems you can open any class.  I don't quite like that.

In F# I was planning this:

```open System.Math```

(note there is no `open static`), and only for static classes (i.e. abstract, sealed in the underlying IL).

The catch 22 is that if we go with the F# approach of naked `open` then our autocomplete lists for `open System.` should really contain all static types, which pollutes the list a little but not too badly (about 20 extra things, though almost none of which you'd want to open)

If we go with the F# approach of naked `open` PLUS being able to open the static content of anything then the autocomplete list would in theory contain all types, i.e. 400+ items.....

My proposal is that 

1. We only allow `open` on static types
2. We only show auto-complete entries for static types

Do people agree with (1)?  It will however mean that the feature doesn't have exact parity with C#....

**Comment by baronfel** [comment]
I think that's a reasonable step and gives the majority of the benefit overall.  I imagine if the need for opening particular classes with static members is that great, the handling of `open`s could be expanded.

**Comment by dsyme** [comment]
> ... if the need for opening particular classes with static members is that great, the handling of opens could be expanded.

My fear is that we will indeed have to do that, because some important C# DSL API gets shipped without declaring a class `static`.  I don't particularly want to have to reopen this design point....

So an alternative would be

1. We allow open on all types
2. We only show auto-complete entries for static types

but that is also kind of a weird position to end in, since auto-complete is not accurate.

A final option is simply to allow `open static` in F# but I would find it odd and a bit of a misfit.

**Comment by baronfel** [comment]
Going with 'open on all', filtering auto-complete is something that's entirely FCS-based? So in theory could be behind a toggle/config option?  VS/FSAC/etc could ship with that filter-flag set to true and allow folks to opt-in to overloads if necessary.

**Comment by FunctionalFirst** [comment]
> If we go with the F# approach of naked `open` PLUS being able to open the static content of anything then the autocomplete list would in theory contain all types, i.e. 400+ items.....

Does autocomplete currently list all available entries? If so, is virtual scrolling / lazy loading an option? Or perhaps only showing the first 50 or so options, which would require typing additional chars to pare the list down.

**Comment by cartermp** [comment]
What's the rationale for `open static` being odd and a misfit?

Let's not get fancy with completion. Its job should be to just suggest symbols that are in scope; not need to apply special rules atop the language semantics.

**Comment by matthid** [comment]
> What's the rationale for open static being odd and a misfit?

The rationale would indeed be interesting to know. Though I personally would skip the `static` if possible as it is more in-line with the existing feature of opening modules.

> If we go with the F# approach of naked open PLUS being able to open the static content of anything then the autocomplete list would in theory contain all types, i.e. 400+ items.....

Afaics IntelliSense is able to handle that due to its already existing filtering logic:
![image](https://user-images.githubusercontent.com/1866463/58988256-8f1dd880-87e1-11e9-87fa-ad0abc3db0cf.png)
So is it really that much of a problem?

Btw - on this occasion - IntelliSense suggests everything on the first character, even today:

![image](https://user-images.githubusercontent.com/1866463/58988367-d906be80-87e1-11e9-8ea9-3be8ad562832.png)

It doesn't seems to "disturb" anyone too much as nobody reported that afaik. 

@dsyme any particular reason why you think it would be a problem listing all types with at least one static member? Is it about non-VisualStudio editors?

**Comment by dsyme** [comment]
> The rationale would indeed be interesting to know. Though I personally would skip the static if possible as it is more in-line with the existing feature of opening modules.

I'm not totally opposed to `open static`.  

I think my concern is that `open` is used for F# modules (which are obviously a lot like static classes), and that people just will get confused by `open static` for that reason.

**Comment by dsyme** [comment]
> I think my concern is that open is used for F# modules (which are obviously a lot like static classes), and that people just will get confused by open static for that reason.

A quick thought: perhaps `open type System.Math` is actually more idiomatic for F#.  This is partly because `type` is used at the declaration site.... This would also allow `open module FSharp.Core.Something` if we want.

**Comment by vasily-kirichenko** [comment]
I'm for just `open System.Math`. It would be nice if it emitted a warning if the type does not contain any accessible static members.

**Comment by dsyme** [comment]
I've discussed this with @cartermp and we've decided to go with just plain "open", even though the intellisense  lists may end up containing additional elements.

**Comment by matthid** [comment]
Just a random thought: I assume this feature makes `type`s in addition to static members way more viable, maybe even more than we want?

**Comment by cartermp** [comment]
I guess we'll learn over time 🙂

**Comment by dsyme** [comment]
> Just a random thought: I assume this feature makes types in addition to static members way more viable, maybe even more than we want?

Yes, it can be a distortive influence on how API and DSL design is done in both F# and C#.

**Comment by cartermp** [comment]
Closing is this is in and shipping

**Comment by dsyme** [comment]
Reopen as the feature is in preview due to lingering design issues

**Comment by dsyme** [comment]
A design issue has been raised about the behaviour of this feature in the presence of method overload sets combined from multiple different classes, see https://github.com/dotnet/fsharp/issues/7400, copied below

Thanks to @TIHan for bringing this up. Consider the following example:

```fsharp
[<AbstractClass; Sealed>]
type One =
    static member M(x: int) = 1

[<AbstractClass; Sealed>]
type Two =
    static member M(x: float) = 1

open One
open Two

let x = M(1.0)
let y = M(1) // Compile error!
```

The `M` that takes an `int` is shadowed, as if these were functions.

@cartermp writes: However, this feels quite unexpected since these are methods, and methods can be overloaded. And as you'd expect, if both `M` methods were defined in the _same_ static class, you could write code that used both overloads. And since they have unique signatures, it's expected that you could call either. Hence, I consider this behavior to be unexpected and/or a bug and not an explicit design decision.

@dsyme says - Not allowing combination of unrelated method overload sets was a deliberate design decision in the now-preview-version of this feature, but we can discuss further. I notice you've updated the RFC with the intended change which is OK since the feedback appears quite strong that this is unexpected.  Personally, from a design perspective I prefer not combining unrelated overload sets - there is a lot of hell and misuse awaiting future users of libraries that make use of this feature. However I can see why you've queried it, and that from an API/interop perspective it makes sense to align with C# to a greater degree.

**Comment by TIHan** [comment]
More interesting things on current preview's behavior:

```fsharp
module Test =

    let x = 1

    let M () = ()

[<AbstractClass; Sealed>]
type Test () =

    static member M() = ()

    static member M(x: int) = ()

    static member M2() = ()

let x = Test.M()

open Test
```

### `let x = Test.M()`, which `M` is it? It's actually the module's.

This whole bit below is not as a result of open static classes; it's existing behavior but it's important to know.

On qualified access,`Test.M`, `Test` will try to resolve as a module first and see if the dot ident, `.M`, will resolve. If not, it will try the type. Irrespective of order, `Test.M` will always resolve to the module's `M`; also no access to `Test` class' `M` overloads. Module always win. 

This is deliberate, but a comment from name resolution asked to review this behavior:
`// Otherwise modules are searched first. REVIEW: modules and types should be searched together.`
Changing this long-standing behavior is definitely a breaking change. I'm not suggesting it; I'm just pointing it out.

As a side example:
```fsharp
module Test =

    let M () : unit = failwith "test module"

[<AbstractClass; Sealed>]
type Test () =

    static member M() = ()

    static member M2() = Test.M()
```
Calling `Test.M2()` will result in the failwith exception. In this case, it's not possible to access the `Test` type's `M` method from within the type unless you change method's name.

In regards to open static classes, what if we want to open both module and type? Which unqualified `M` will win in that case (will order matter?)? Well, we first need to ask "how" we can open both.

### `open Test` , what do you think it opens? With open static classes enabled, it opens *both*.

The current preview behavior opens both.

I now have access to `x`, `M`, and `M2`. `x` and `M` are from the `Test` module and `M2` is from the `Test` class.

This is a breaking change. It means more items could be available in scope for someone who happens to have a static class of the same module name in scope; same goes with namespaces. These items have the potential to shadow any existing items before it; therefore, is a breaking change.

Ex:
```fsharp
let X = 1

[<AbstractClass; Sealed>]
type System () =

    static member X = "doot"

open System

let hopac : Int32 = X
```
Code will compile in 4.6.
Code will not compile with open static classes, errors on `let hopac : Int32 = X` saying that `X` type mismatches on string and int.

I sort of figure this might be a bug, but I'm going to assume everything is deliberate. In fact, I think it is deliberate because the same thing would happen for a `module System`.

### Suggested Behavior

1. `open Test` should not open both. It should only open the module and at least have the module or namespace always take precedence over types in scope (otherwise we will have a breaking change) 
    - In the example, if we allowed `open Test` to choose the type, means any existing code that relied on the module opening will certainly fail.

2. Assuming that you can't change the name of either module or type, how can you open both the module and type?
    - You could do a type abbreviation `type TestType = Test; open TestType` (though you can't open type abbrevs right now (bug?)). Assuming we allow opening types that are abbreviated, this doesn't feel natural or good to do. This should not be the approach we take.
    - In some scenarios, you could be fully qualified with a namespace that resolves the ambiguity, but what if the type and module live on the same namespace? You still have the problem.
    - The only other option is to simply denote that you are explicitly opening a type, `open type Test` / `open static Test`. Personally, I like `open type Test`. If an analogous `open module Test` were to be introduced by this, I would vote against that as we would have two ways to do the exact same thing with opening a module. **Update**: However, I found that opening a module could also open the namespace of the same name. Means `open module` could technically have a use, to prevent those namespaces from being opened. Also, what about `open namespace` to prevent opening modules? Just food for thought.

3. Assuming that you can open both module and type, `open Test; open type Test`, will order matter?
    - I think order should be significant; it should shadow items like you would normally think. Basically, ignoring the weird behavior with the qualified names, `Test.M`.

**Comment by dsyme** [comment]
Cool, thanks for looking into these examples.  It's crucial to get them under test regadless of the resolution we choose.

> Changing this long-standing behavior is definitely a breaking change. I'm not suggesting it; I'm just pointing it out.

Right, agreed shouldn't change this.  We could add a warning when the resolution would differ by searching the most-recently-in-scope first? Indeed we could add an optional warning for any ambiguous lookups, ala C#, I'd be happy with that.

> This is a breaking change. It means more items could be available in scope for someone who happens to have a static class of the same module name in scope; same goes with namespaces. These items have the potential to shadow any existing items before it; therefore, is a breaking change.

I'm not really concerned by this breaking change if `open` is restricted to opening static classes.  Very little code declare an `AbstractClass; Sealed` in F# today.  Static classes in .NET are a little more common but I still don't think this is a significant risk of breaking real world code.

If `open C` also opens non-static classes then we have more of an issue and I do think we need to switch to use `open type C` for all opening of static classes.  Perhaps that's a much easier way to resolve this issue.  It would also help with `open type C<int>`

**Comment by cartermp** [comment]
I mostly agree with @TIHan here, though I'm leery about shadowing given that we're not working strictly with module-bound values, but with both module-bound values and methods. I think that if there's an ambiguity in names, the module should always win given that this happens today if you fully qualify. In both cases, if you wanted the member, the solution is the same: rename something. If you had to rename something in one case but had to reorder `open` declarations in another, that'd be inconsistent.

I think we should go with `open type C` (or some other keyword if we're not hot on `type`) given that the possible extension to opening any type with static members on it is relevant and valuable from an interop standpoint. And if it helps with opening generic static classes, sweet!

**Comment by dsyme** [comment]
> I think we should go with open type C (or some other keyword if we're not hot on type) given that the possible extension to opening any type with static members on it is relevant and valuable from an interop standpoint. And if it helps with opening generic static classes, sweet!

OK, so this is the POR to solve this issue?

**Comment by cartermp** [comment]
I think so, yes. @tihan?

**Comment by eiriktsarpalis** [comment]
> I think we should go with open type C (or some other keyword if we're not hot on type)

How about `open static C`? Would be more on par with C#.

**Comment by TIHan** [comment]
`open type` seems like the right way to go. `type` feels natural given we have `namespace` and `module` at similar declaration levels.

`open static` may also be fine. Since this is still going to be in preview, we can change it if we feel that `open static` might be more appropriate. 

One argument against `open static` is: `using static` in C#, not only allows unqualified access to static members, but it also allows access to nested types. In that sense, you are basically opening the type for more than just static members.

A rebuttal against that argument is: in F#, you can't even create nested types; you are only going to have static members in F# defined classes. Also, it is probably more common to want to access static members rather than nested types.

Let's choose `open type` for preview and see how it feels. Me personally, I'm open to either. I just think `type` fits at the moment.

---

As I mentioned above, C# also allows access to nested types as unqualified in `using static`. I don't know if we should do the same but only for .NET types?

---

Regarding @cartermp 's comment on the shadowing of module values and type static members, 
> I think that if there's an ambiguity in names, the module should always win given that this happens today if you fully qualify.

The reason it chooses module value is because it tries to find the module name first; so it's not exactly choosing the value over the member. It's the resolution on the qualified module/type name. You also have a similar behavior with types within modules:
```fsharp
module Test =

    type M() = class end

type Test() =

    static member M() = 1

let x = Test.M()
```
`Test.M()` is the constructor for the type `M` in module `Test`. 

Applying the ambiguity module/type-name behavior, would mean values and types always win against members, fields, properties, events.
Ex:
```fsharp
let X = 1

[<AbstractClass; Sealed>]
type Test () =

    static member X = "hi"

open Test

let x = X
```
`x` wouldn't be a `string`, it would be `int` referring to the `let X` at the beginning.

Not sure how I feel. The next example below makes me worried:

```fsharp
module Mod =

    type Test () =

        static member M() = 1

        static member M2() = 1

open Mod

type Test () =

    static member M(x: int) = x

let x = Test.M()
```
`Test.M()` will give me an error; I can only get access to `M(int)`, but I can get access to `M2()`. This is due to name resolution looking at the types from recent open order when qualified with `Test`, then finding a method group `M` and stops.

By the logic we want to remain consistent with the ambiguity behavior of module/type names when fully qualified, overloading would be out of the question when opening multiple types; was a major factor for re-opening this discussion. Also means the current preview is almost consistent with the behavior.

I think we should treat fully qualified behavior separate from opening item behavior; they are just two separate things. We should design the rules on opening types/static classes that are most useful and expected to the user. The one thing we should be consistent about is when we open a single type-name in which the type-name is ambiguous with other types, you should get the same static members that you would see if you were going to fully qualify with the type. It's only when we open multiple different types with different names do we apply a new set of rules, allowing overloading, method group priority, etc.

**Comment by dsyme** [comment]
> As I mentioned above, C# also allows access to nested types as unqualified in using static. I don't know if we should do the same but only for .NET types?

Yes, we should - great catch, I'll add it to the RFC resolved-but-not-implemented issues

**Comment by dsyme** [comment]
I've submitted an update to the RFC to adopt `open type` as well as the other design issues https://github.com/fsharp/fslang-design/pull/389

**Comment by dsyme** [comment]
> How about open static C? Would be more on par with C#.

I also agree with `open type` for F#.  I know it's a small thing but it does allow for `open module XYZ` or `open val <expr>`

**Comment by abelbraaksma** [comment]
Looks promising! I read this thread and the language design, but I feel like there are perhaps still a few minor issues or ambiguities with it that may need addressing:

1. If `B` with method `foo` is nested in class `A`, and we have `open type A`, can we use both `B.foo()` and `foo()` or only the latter?
2. Near the end, the resolution order of a method, property etc is detailed. Step 8 says it recursively resolves nested classes. Will this be depth-first or depth-last traversal, or any other order? While perhaps a fringe case, the resolution order should be deterministic.
3. If `B` is an alias of type `A`, can I do `open type B` or is this illegal?
4. In the 1-8 steps of order resolution, constants are not mentioned. Are these considered equal to fields?
5. While not made explicit in this proposal, extension methods in a different namespace than the opened type would need to be opened prior to `open type` to be in scope for proper resolution, right?
6. There's a section on generic classes, but nothing on generic methods. F# would normally auto-infer generics. Will such generics be supported and if so, in terms of resolution, when are they resolved?

Example of the latter:

```f#
[<AbstractClass; Sealed>]
type Example1 =
    static member Return<'T> (value: 'T) = value

[<AbstractClass; Sealed>]
type Example2 =
    static member Return (value: int) = value

open type Example1
open type Example2

Return 12   // will this call Example1 or Example2 or raise en error?
```

**Comment by chkn** [comment]
Why not both? We could extend unqualified `open` to include static classes, but preferring modules/namespaces so as not to be a breaking change. In addition, we allow `open type` to open any class, matching the C# behavior

I say this because I agree with @dsyme's comment from above:

> I think my concern is that open is used for F# modules (which are obviously a lot like static classes), and that people just will get confused by open static for that reason.

As a library author, I could imagine a case where you'd need to use methods in a static class, which can behave a lot like functions if they only take one argument, and not want consumers of the API to be exposed to this implementation detail (by requiring them to do `open type` instead of just `open`)

**Comment by dsyme** [comment]
Hi, just to say @TIHan and @cartermp are doing a (hopefully final) design reiew of this feature this afternoon

If you have comments or particular points please feel free to remind us of them here

**Comment by abelbraaksma** [comment]
While I posted this September last year, I think these still need addressing (or I've missed the update): https://github.com/fsharp/fslang-design/issues/352#issuecomment-533705040

**Comment by dsyme** [comment]
@abelbraaksma Thanks.  I'll update the RFC and we'll aim to revolve these this afternoon

**Comment by dsyme** [comment]
> If B with method foo is nested in class A, and we have open type A, can we use both B.foo() and foo() or only the latter?

If `B` is a nested .NET type then you must use `B.foo` unless you explicitly 

TODO: check we have testing for this.

> Near the end, the resolution order of a method, property etc is detailed. Step 8 says it recursively resolves nested classes. Will this be depth-first or depth-last traversal, or any other order? While perhaps a fringe case, the resolution order should be deterministic.

Depth first.

> If B is an alias of type A, can I do open type B or is this illegal?

Yes you can 

TODO: check we have testing for this:

> In the 1-8 steps of order resolution, constants are not mentioned. Are these considered equal to fields?

As in literal constants, e.g. enums?  Yes they are like fields.  

> While not made explicit in this proposal, extension methods in a different namespace than the opened type would need to be opened prior to open type to be in scope for proper resolution, right?

Yes

TODO: check we have testing for this

> There's a section on generic classes, but nothing on generic methods. F# would normally auto-infer generics. Will such generics be supported and if so, in terms of resolution, when are they resolved?

See below

**Comment by krauthaufen** [comment]
Hey, although I think the feature is really valuable in several situations, I'd like to ask a few questions:
1. Isn't this conceptually identical to allowing module-function overloads?
2. If so, will the overload resolution for opened static classes "contain" the current module-function-resolution?

I'm not suggesting we allow module-overloads but I think much of F#'s attractiveness originates in its simplicity and overloaded pseudo-global functions may make things hard for beginners (especially when libraries start using that a lot).

Nonetheless I think the feature would simplify lots of things where overloaded functions are realized using SRTP and showing quite unhelpful type signatures

**Comment by dsyme** [comment]
Regarding this example from above by @abelbraaksma, this will call Example2:

```fsharp
[<AbstractClass; Sealed>]
type Example1 =
    static member Return<'T> (value: 'T) = value

[<AbstractClass; Sealed>]
type Example2 =
    static member Return (value: int) = value

open type Example1
open type Example2

Return 12   // will this call Example1 or Example2 or raise en error?
```

**Comment by dsyme** [comment]
> Isn't this conceptually identical to allowing module-function overloads?

There's a relation but in practice a long way from identical:

1. The fact that method overload sets are not combined means the ability to add new reslutions is not open.  This means the overloads have to be designed to work together

2. The fact that object programming syntax has to be used to make the definitions in F# means that the process falls into the part of the language where more design/experience is needed 

3. What you can do with Module/value/function programming is not disturbed if you don't use this feature.

> If so, will the overload resolution for opened static classes "contain" the current module-function-resolution?

No, if a static class contains an overload set for `M` then that shadows any module-value-function resolution named `M`

**Comment by dsyme** [comment]
@cartermpm @TIHan and I spent an hour discussing this. We decided

1. We will intentionally stick with the decision *not* to combine method overload sets from unrelated types.  This is for two reasons - software engineering and the interaction with redefinition of types in F# Interactive and F# Notebooks. 

   The long term software engineering problems with allowing open combiantion can be quite extreme and there are good reasons not to allow it.  C# allows it, and it is possible that some C# APIs will be designed with the intention of allowing open combination.  These will be more difficult to use from F# as they may require qualification or a facade that combines and resolves the problems.

2. We confirmed we would change the syntax to `open type`

3. We confirmed we would allow open on non-static types

**Comment by dsyme** [comment]
@TIHan note we must also do this: 

> Note: In the preview nested types are not revealed when opening a static type

**Comment by TIHan** [comment]
Makes sense, we will stop putting the nested types into scope.

**Comment by dsyme** [comment]
> Makes sense, we will stop putting the nested types into scope.

@TIHan We **should** bring them into scope.  I think "preview" is **not** doing this :-)

**Comment by dsyme** [comment]
@TIHan The RFC also needs to be udpated to make it clear that method overload sets are not being combined.  I think we adjsuted it to combine them and now need to revert that.

**Comment by abelbraaksma** [comment]
> Regarding this example from above by @abelbraaksma, this will call Example2:

Thanks, that clarifies the resolution for generic members. Iirc this is the same rule as for other stuff in F#: last one wins.

**Comment by abelbraaksma** [comment]
@dsyme, we can presently extend a type locally (no warning) or elsewhere (warning). Will static types be following the same rules, or will extensions be prohibited?

**Comment by TIHan** [comment]
> @TIHan We should bring them into scope. I think "preview" is not doing this :-)

Ah, oops. I read it backwards :) - Ok, that shouldn't be a problem.

**Comment by TIHan** [comment]
> I think we adjsuted it to combine them and now need to revert that.

I don't think we adjusted combining them yet. I had a PR that did allow it, but now we don't need it.

**Comment by dsyme** [comment]
> @dsyme, we can presently extend a type locally (no warning) or elsewhere (warning). Will static types be following the same rules, or will extensions be prohibited?

Extensions are permitted

**Comment by dsyme** [comment]
Resolution of three design issues:

1. Consider  "open type Namespace.UnionType" and "open type Namespace.RecordType" where 

   * "UnionType"/"RecordType" is not RequireQualifiedAccess
    * "Namespace" has not been opened

   For example consider https://github.com/fsprojects/FSharp.Formatting/blob/master/src/FSharp.Formatting.Common/HtmlModel.fs#L4 then 

   ```fsharp
   open  type FSharp.Formatting.HtmlModel.HtmlProperties
   ```

   **Resolution:** Union cases and record field labels should count as static content when processing `open type` and thus brought into scope..

2. Opening types that contain operator definitions, e.g. 

   ```fsharp
   type C() = 

       // What about this?  Is this static content in C#??  
       static member (+) (a,b) = ...

   open type C
   ```

   This is bad as the name `op_Addition` is brought into scope, but it wipes out the FSharp.Core.Operators definition of `+` (known to F# as `op_Addition`).  This means `1+1` no longer resolver. 

   In C#, the operator is never directly callable and while `op_Addition` is brought into scope, the resolution of `1+1` is not changed.

   **Resolution:** in F# we should omit all static members with an operator name (beginning with `op_`) from static content when processing `open type`.


3. Opening types that contain C# style extension members

   **Resolution:** In this case, extension members are brought into scope

**Comment by gusty** [comment]
Regarding 2. isn't the same case with global functions?

I mean, if the type defined a `sqrt` member, wouldn't our "global" `sqrt` be shadowed and stop working?

I understand it is bad, but it might be that the whole feature is bad in general and it should be used carefully.

**Comment by cartermp** [comment]
@gusty True, I don't quite understand the resolution since it's no different than any component that does this:

```fsharp
module Beef =
    let inline ( + ) x y = x - y
    let inline ( - ) (x, y) = x = y
    
open Beef
    
    
printfn "%d" (1 + 2) // prints '-1' because '+' is resolved from the module
1 - 2 // error, doesn't resolve that way
```

@dsyme @TIHan why treat operators different just for this feature?

**Comment by cartermp** [comment]
FWIW I do not believe this quibble should keep us from merging this so we can flight it in previews

**Comment by dsyme** [comment]
@cartermp @gusty The huge danger is that C# types with operators are designed for opening, but take no consideration that simply adding a `+` operator will completely nuke the `+` operator of F# (where it is completely irrelevant in the C# version of the opening feature).

If the C# feature isn't affecting the resolution of `+` then I don't think the F# import of that feature should nuke `1+1`.

**Comment by abelbraaksma** [comment]
> Resolution: Union cases and record field labels should count as static content when processing open type and thus brought into scope..

I don't understand this: the example seems to talk about numeric unions, aka, enums, but the text, also in the rfc, seems to refer to F# unions, aka DU's (note that the rfc has an error in the examples here: it shows `open`, not `open type`). 

I can see why we bring DU cases into scope, but I don't see a good reason for bringing CLI enum fields into scope: it's long been a tradition that enums (not DU) must use qualified access. I fear that if we change this, it could cause compatibility issues, or confusion, if both a DU (no qualified access) and an enum (previously qualified access) are brought into scope.

I propose to continue requiring qualified access for CLI enums, and only bring DU cases into scope if `open type` refers to a DU. For CLI enums, it could still bring static methods into scope, not sure on that; just not the fields.
